src/translation.cpp

```diff
@@ -145,11 +211,337 @@ void Translations::loadTranslation(const std::string &data)
+			case L'0': case L'1': case L'2': case L'3': case L'4': case L'5': case L'6': case L'7': {
+				size_t j = 0;
+				wchar_t c = 0;
+				for (; j < 3 && i+j < str.length() && L'0' <= str[i+j] && str[i+j] <= L'7'; j++) {
+					c = c * 8 + (str[i+j] - L'0');
+				}
+				if (c <= 0xff) {
+					result.push_back(c);
+				}
+				i += j;
+				continue;
+			}
+			case L'x': {
+				i++;
+				if (i >= str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				size_t j = 0;
+				unsigned char v;
+				for (; i+j < str.length() && hex_digit_decode((char)str[i+j], v); j++) {
+					c = c * 16 + v;
+				}
+				if (j == 0) {
+					errorstream << "Invalid escape sequence \\x, ignoring" << std::endl;
+					continue;
+				}
+				// If character fits in 16 bits and is not part of surrogate pair, insert it.
+				// Otherwise, silently drop it: this is valid since \x escape sequences with
+				// values above 0xff are implementation-defined
+				if ((c < 0xd800) || (0xe000 <= c && c <= 0xffff)) {
+					result.push_back(c);
+				}
+				i += j;
+				continue;
+			}
+			case L'u': {
+				i++;
+				if (i + 4 > str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char16_t c = 0;
+				bool ok = true;
+				for (size_t j = 0; j < 4; j++) {
+					unsigned char v;
+					if (str[i+j] <= 0xff && hex_digit_decode((char)str[i+j], v)) {
+						c = c * 16 + v;
+					} else {
+						errorstream << "Invalid unicode escape sequence \"\\u" << wide_to_utf8(str.substr(i, 4)) << "\", ignoring" << std::endl;
+						ok = false;
+						break;
+					}
+				}
+				if (ok) {
+					wide_add_codepoint(result, c);
+				}
+				i += 4;
+				continue;
+			}
+			case L'U': {
+				i++;
+				if (i + 8 > str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				bool ok = true;
+				for (size_t j = 0; j < 8; j++) {
+					unsigned char v;
+					if (str[i+j] <= 0xff && hex_digit_decode((char)str[i+j], v)) {
+						c = c * 16 + v;
+					} else {
+						errorstream << "Invalid unicode escape sequence \"\\U" << wide_to_utf8(str.substr(i, 8)) << "\", ignoring" << std::endl;
+						ok = false;
+						break;
+					}
+				}
+				if (ok) {
+					wide_add_codepoint(result, c);
+				}
+				i += 8;
+				continue;
+			}
+			default: {
+				errorstream << "Unknown escape sequence \"\\" << str[i] << "\", ignoring" << std::endl;
+				break;
+			}
+		}
+		i++;
+	}
+	return result;
+}
+
+void Translations::loadPoEntry(const std::map<std::wstring, std::wstring> &entry) {
+	// Process an entry from a PO file and add it to the translation table
+	// Assumes that entry[L"msgid"] is always defined
+	std::wstring textdomain;
+	auto ctx = entry.find(L"msgctxt");
+	if (ctx != entry.end()) {
+		textdomain = ctx->second;
+	} else {
+		textdomain = L"";
```

Comment:
Since other PO file generators do not necessarily generate `msgctxt` entries it would IMO be a good idea to use the base name of the file (i.e. without the `LANG.po` suffix) as the textdomain if a `msgctxt` entry is not available.

Reply:
Done